### PR TITLE
feature #47: binarizeEdgePlus

### LINKED ISSUE #47: [Feature request] binarizeEdgePlus
Hi @vigri .

```cpp
/**
 * \brief Image binarization using EdgePlus local/global thresholding method.
 *
 * EdgePlus, zvezdochiot 2023. "Adaptive/global document image binarization".
 */
BinaryImage binarizeEdgePlus(const QImage& src, QSize windowSize, double k = 0.34);
```

```cpp
BinaryImage binarizeEdgePlus(const QImage& src, const QSize windowSize, const double k) {
  if (windowSize.isEmpty()) {
    throw std::invalid_argument("binarizeSauvola: invalid windowSize");
  }

  if (src.isNull()) {
    return BinaryImage();
  }

  QImage gray(toGrayscale(src));
  const int w = gray.width();
  const int h = gray.height();

  IntegralImage<uint32_t> integralImage(w, h);
 
  uint8_t* grayLine = gray.bits();
  const int grayBpl = gray.bytesPerLine();

  for (int y = 0; y < h; ++y, grayLine += grayBpl) {
    integralImage.beginRow();
    for (int x = 0; x < w; ++x) {
      const uint32_t pixel = grayLine[x];
      integralImage.push(pixel);
    }
  }

  const int windowLowerHalf = windowSize.height() >> 1;
  const int windowUpperHalf = windowSize.height() - windowLowerHalf;
  const int windowLeftHalf = windowSize.width() >> 1;
  const int windowRightHalf = windowSize.width() - windowLeftHalf;

  grayLine = gray.bits();
  for (int y = 0; y < h; ++y) {
    const int top = std::max(0, y - windowLowerHalf);
    const int bottom = std::min(h, y + windowUpperHalf);  // exclusive
    for (int x = 0; x < w; ++x) {
      const int left = std::max(0, x - windowLeftHalf);
      const int right = std::min(w, x + windowRightHalf);  // exclusive
      const int area = (bottom - top) * (right - left);
      assert(area > 0);  // because windowSize > 0 and w > 0 and h > 0
      const QRect rect(left, top, right - left, bottom - top);
      const double windowSum = integralImage.sum(rect);
 
      const double rArea = 1.0 / area;
      const double mean = windowSum * rArea;
      const double origin = grayLine[x];
      // edge = I / blur (shift = -0.5) {0.0 .. >1.0}, mean value = 0.5
      const double edge = (origin + 1) / (mean + 1)  - 0.5;
      // edgeplus = I * edge, mean value = 0.5 * mean(I)
      const double edgeplus = origin * edge;
      // return k * edgeplus + (1 - k) * I
      double retval = k * edgeplus + (1.0 - k) * origin;
      // trim value {0..255}
      retval = (retval < 0.0) ? 0.0 : (retval < 255.0) ? retval : 255.0;
      grayLine[x] = (int)retval;
    }
    grayLine += grayBpl;
  }
  return BinaryImage(src, BinaryThreshold::otsuThreshold(gray));
}  // binarizeEdgePlus
```
See also:
* https://github.com/4lex4/scantailor-advanced/issues/196
* https://github.com/4lex4/scantailor-advanced/issues/197
* https://github.com/ScanTailor-Advanced/scantailor-advanced/issues/8

### PR DESCRIPTION
**Pull request by zvezdochiot**
Hi @vigri and @lightsilverberryfox .

:warning: This PR is just a recipe. I was only able to build `binarizeEdgePlus` into the general functionality. At the same time, I use the Sauvola options, which I don’t think is correct (the new threshold requires smaller area values ​​and larger coefficient values). Due to this, projects with a new threshold are safely saved. Use this PR as a recipe.

Good Luck.

---

## Example

| Origin  | Otsu threshold |  Sauvola threshold | Wolf threshold | EdgePlus threshold | EdgePlus threshold | EdgePlus threshold |
| --- | --- | --- | --- | --- | --- | --- |
| (screenshot from 600dpi scan) | (default) | (default: ws:200 c:0.34) | (default) | (default: ws:200 c:0.34) | (manual: ws:20 c:0.80) | (manual: ws:5 c:0.99 Equalize:0ff Savitzky-Golay:off) |
| screen-origin | screen-otsu | screen-sauvola | screen-wolf | screen-edgeplus-default | screen-edgeplus-a20-c80 | screen-edgeplus-a5-c99-off-sav-norm |

## Links

* [Build this PR](https://github.com/ImageProcessing-ElectronicPublications/scantailor-advanced/releases/tag/v1.0.18)
* [OCR forum (RU): ScanTailor Advanced; ideas for improving the clarity of scans](http://publ.lib.ru/cgi/forum/YaBB.pl?num=1683219756)
* [Scan Tailor (part 2)](http://forum.ru-board.com/topic.cgi?forum=5&topic=32945)

Closes #47

### CODE REVIEW

**Comment by zvezdochiot** [comment]
Hi @vigri , @lightsilverberryfox .

The usefulness of the new thresholds has been confirmed:
| Origin | Otsu | Sauvola | Wolf | EdgePlus | BlurDiv |
| --- | --- | --- | --- | --- | --- |
| (600dpi) | (default) | (default) | (default) | (ws:20,c:0.75) | (ws:20,c:0.75) |
| test | test otsu | test sauvola | test wolf | test edgeplus_20_075 | test blurdiv_20_075 |
See threshold `BlurDiv`.

**Comment by zvezdochiot** [comment]
#59

**Comment by zvezdochiot** [comment]
Mod **epdbta** of release 1.0.18: https://github.com/ImageProcessing-ElectronicPublications/scantailor-advanced/releases/tag/v1.0.18

**Comment by alfish2000** [comment]
Hi!
Is there any chance that the PRs here get merged and e.g. a new version 1.0.19 is released?
It gets confusing because it seems that this PR is now released here https://github.com/ImageProcessing-ElectronicPublications/scantailor-advanced as 1.0.18

**Comment by zvezdochiot** [comment]
@alfish2000 say:
>Is there any chance that the PRs here get merged and e.g. a new version 1.0.19 is released?

I don't hope so. Therefore, I release mods for existing releases. ;)

**Review by lightsilverberryfox — COMMENTED** [review]
Okay, the code seems to be fine, I don't understand how the EdgePlus algorithm works though... But if it works it should be okay. Let's see if @vigri has something to say.

**Comment by mara004** [comment]
>> Is there any chance that the PRs here get merged and e.g. a new version 1.0.19 is released?

> I don't hope so. Therefore, I release mods for existing releases. ;)

I'm confused by your "I don't hope so."
Are you still OK with this being merged and included in a release here eventually?
Otherwise, I think you should close the PR, mark it as draft, or similar.

**Comment by zvezdochiot** [comment]
Hi @mara004 .

I don't have enough experience with Qt. If there are bugs, it will be difficult for me to eliminate them. If this PR enters the mainstream, I will be happy. Doubts are caused purely by lack of skill.

**Comment by zvezdochiot** [comment]
Hi @mara004 , @lightsilverberryfox , @vigri .

Demo of new thresholds:
| Origin | EdgePlus |
| --- | --- |
| 150dpi | 600dpi (d:16, ws: 20, c: 0.99) |
| p-004 | 
p-004 pbm |

**Comment by zvezdochiot** [comment]
Hi @mara004 , @lightsilverberryfox , @vigri .

Github Action: [build-linux-qt5](https://github.com/ImageProcessing-ElectronicPublications/scantailor-advanced/actions/runs/5534117653/jobs/10098638821#logs)
```log
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libi/libinput/libinput-bin_1.20.0-1ubuntu0.2_amd64.deb  404  Not Found [IP: 52.147.219.192 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libi/libinput/libinput10_1.20.0-1ubuntu0.2_amd64.deb  404  Not Found [IP: 52.147.219.192 80]
```

**Review by lightsilverberryfox — APPROVED** [review]
Okay, @zvezdochiot 

After some testing and reviewing, I believe the PR is ready to be merged, I didn't encounter any issues and this could be a very useful feature, I would normally wait for @vigri to offer their review, however he hasn't been here for three months already, so, I don't know what to do.

If you believe it's ready to be merged, I'll merge it. I can tell you put a lot of effort in this PR and wanted to say thank you, you have also offered your input on other issues and bugs, so if vigri sees this, he may want to consider sending you an invitation to our organization so you have write access (I can't invite you myself, I don't have owner permissions).

File: src/core/filters/output/BlackWhiteOptions.cpp
Comment (vigri) [inline src/core/filters/output/BlackWhiteOptions.cpp:75]: What do you think about refactoring this one to a switch/case? Just for readability?
Just like in `QString BlackWhiteOptions::formatBinarizationMethod(BinarizationMethod type) `

File: src/dewarping/DetectVertContentBounds.cpp
Comment (vigri) [inline src/dewarping/DetectVertContentBounds.cpp:250]: Is `<=9` instrad of `<14` really intended here?

**Review by vigri — CHANGES_REQUESTED** [review]
Overall I see no problems. Please see my comments. Especially the one with the qt-version.

**Comment by vigri** [comment]
> I can tell you put a lot of effort in this PR and wanted to say thank you, you have also offered your input on other issues and bugs, so if vigri sees this, he may want to consider sending you an invitation to our organization so you have write access (I can't invite you myself, I don't have owner permissions).

apologies very much.

Thank you very much @zvezdochiot.
As @lightsilverberryfox said: If you are interested, I can send you an invitation. Please let me know.

Regarding this PR: Pleasse see my "review".

File: src/dewarping/DetectVertContentBounds.cpp
Comment (zvezdochiot) [reply src/dewarping/DetectVertContentBounds.cpp:250]: Hi @vigri .

@plzombie specifically pointed out just such a version in the given condition.

File: src/core/filters/output/BlackWhiteOptions.cpp
Comment (zvezdochiot) [reply src/core/filters/output/BlackWhiteOptions.cpp:75]: Hi @vigri .

`switch/case` doesn't work on strings.

**Comment by zvezdochiot** [comment]
@lightsilverberryfox say:
>Is issue #59 also solved by this PR?

This PR closes the issues:
* #47 
* #59 
*

File: src/core/filters/output/BlackWhiteOptions.cpp
Comment (vigri) [reply src/core/filters/output/BlackWhiteOptions.cpp:75]: Sorry, my fault. You are right, we are here in cpp :)

**Review by vigri — APPROVED** [review]
Thank you very much again for this PR!

**Comment by mara004** [comment]
> PS: Me, @plzombie and @trufanov-nok will oversee the STA fork at organization https://github.com/ImageProcessing-ElectronicPublications. Participation in https://github.com/ScanTailor-Advanced is not required.

Wouldn't it be better to join efforts in one repo/organization, instead of all the fragmentation?

**Comment by zvezdochiot** [comment]
@mara004 say:
>Wouldn't it be better to join efforts in one repo/organization

Maybe, but not now. I myself specialize in console utilities (GUI is not mine), @plzombie is busy with [depress](https://github.com/plzombie/depress) (it is more important to him), @trufanov-nok is in charge of [STU](https://github.com/trufanov-nok/scantailor-universal). Everyone is busy with their projects.

PS: Until I figure out the STA (XML) projects, my help is over. I won't figure it out until I see how it's done. Such are the things.

**Comment by zvezdochiot** [comment]
Hi @mara004 , @lightsilverberryfox , @vigri .

New version?
1.0.19?
I need for forums.

**Comment by vigri** [comment]
> Hi @mara004 , @lightsilverberryfox , @vigri .
> 
> New version? 1.0.19? I need for forums.
> 
> Current:
> 
> https://github.com/ScanTailor-Advanced/scantailor-advanced/blob/b0788fc5055ab8f7f7a6b5d727dd13eea1006212/version.h.in#L6

I've uploaded 1.0.19 now.
https://github.com/ScanTailor-Advanced/scantailor-advanced/releases/tag/v1.0.19

**Comment by ireun** [comment]
Guys it would be great to have the docs improves to explain the new thresholds, maybe with examples or when to use what. :)

**Pull request merged**
